### PR TITLE
[EOSF-768]  Ember osf web

### PR DESCRIPTION
**Pull request by hmoco**
## Purpose
Add the docker-compose, settings, routes, and views needed to use the new ember-osf-web app to serve the ember pages we will be incorporating, starting with quick files.

## Changes
Routes for new pages that go under ember-osf-web are added to array in `ember_osf_web/views.py`.

## Ticket

https://openscience.atlassian.net/browse/EOSF-768

### CODE REVIEW

File: website/views.py [hunk 1]
```diff
@@ -303,6 +304,9 @@ def resolve_guid(guid, suffix=None):
 
             return send_from_directory(preprints_dir, 'index.html')
 
+        if isinstance(referent, BaseFileNode) and referent.is_file and referent.node.is_quickfiles:
+            return send_from_directory(quickfiles_dir, 'index.html')
+
```
Comment (jamescdavis) [inline website/views.py]: Also need to add `if PROXY_EMBER_APPS:` similar to preprints above

File: website/quickfiles/views.py [hunk 2]
```diff
@@ -0,0 +1,9 @@
+# -*- coding: utf-8 -*-
+import os
+from flask import send_from_directory
+from website.settings import EXTERNAL_EMBER_APPS
+
+quickfiles_dir = os.path.abspath(os.path.join(os.getcwd(), EXTERNAL_EMBER_APPS['quickfiles']['path']))
+
+def use_ember_app(**kwargs):
```
Comment (jamescdavis) [inline website/quickfiles/views.py]: Also need to add `if PROXY_EMBER_APPS:` here

Comment (hmoco) [reply website/quickfiles/views.py]: ah

File: website/routes.py [hunk 3]
```diff
@@ -43,6 +43,7 @@
 from website.discovery import views as discovery_views
 from website.conferences import views as conference_views
 from website.preprints import views as preprint_views
+from website.quickfiles import views as quickfiles_views
```
Comment (jamescdavis) [inline website/routes.py]: s/quickfiles/ember_osf_web/g

File: website/routes.py [hunk 4]
```diff
@@ -802,6 +803,12 @@ def make_url_map(app):
             OsfWebRenderer('profile/personal_tokens_detail.mako', trust=False)
         ),
 
+        Rule(
+            '/<uid>/quickfiles/',
```
Comment (jamescdavis) [inline website/routes.py]: Use this a list defined in ember_osf_web_views, e.g. `ember_osf_web_views.routes`

File: website/settings/local-dist.py [hunk 5]
```diff
@@ -40,6 +40,11 @@
         'url': '/registries/',
         'server': 'http://192.168.168.167:4300',
         'path': '/registries/'
+    },
+    'quickfiles': {
```
Comment (jamescdavis) [inline website/settings/local-dist.py]: s/quickfiles/ember_osf_web/

File: website/settings/local-travis.py [hunk 6]
```diff
@@ -31,6 +31,11 @@
         'url': '/preprints/',
         'server': 'http://localhost:4200',
         'path': os.environ.get('HOME') + '/preprints/'
+    },
+    'quickfiles': {
```
Comment (jamescdavis) [inline website/settings/local-travis.py]: s/quickfiles/ember_osf_web/

File: website/views.py [hunk 7]
```diff
@@ -29,6 +29,7 @@
 
 logger = logging.getLogger(__name__)
 preprints_dir = os.path.abspath(os.path.join(os.getcwd(), EXTERNAL_EMBER_APPS['preprints']['path']))
+quickfiles_dir = os.path.abspath(os.path.join(os.getcwd(), EXTERNAL_EMBER_APPS['quickfiles']['path']))
```
Comment (jamescdavis) [inline website/views.py]: s/quickfiles/ember_osf_web/g

File: website/views.py [hunk 8]
```diff
@@ -303,6 +304,13 @@ def resolve_guid(guid, suffix=None):
 
             return send_from_directory(preprints_dir, 'index.html')
 
+        if isinstance(referent, BaseFileNode) and referent.is_file and referent.node.is_quickfiles:
+            if PROXY_EMBER_APPS:
+                resp = requests.get(EXTERNAL_EMBER_APPS['quickfiles']['server'], stream=True)
```
Comment (jamescdavis) [inline website/views.py]: s/quickfiles/ember_osf_web/

File: website/views.py [hunk 9]
```diff
@@ -303,6 +304,13 @@ def resolve_guid(guid, suffix=None):
...
+                return Response(stream_with_context(resp.iter_content()), resp.status_code)
+
+            return send_from_directory(quickfiles_dir, 'index.html')
```
Comment (jamescdavis) [inline website/views.py]: s/quickfiles/ember_osf_web/

File: website/routes.py [hunk 10]
```diff
@@ -802,6 +803,12 @@ def make_url_map(app):
...
+            'get',
+            quickfiles_views.use_ember_app,
```
Comment (jamescdavis) [inline website/routes.py]: s/quickfiles/ember_osf_web/

File: website/quickfiles/views.py [hunk 11]
```diff
@@ -0,0 +1,14 @@
+# -*- coding: utf-8 -*-
+import os
+import requests
+from flask import send_from_directory, Response, stream_with_context
+
+from website.settings import EXTERNAL_EMBER_APPS, PROXY_EMBER_APPS
+
```
Comment (jamescdavis) [inline website/quickfiles/views.py]: Add routes list, e.g.
```
routes = [
    '/quickfiles/',
    '/<uid>/quickfiles/'
]
```
This way we can just append to this list as OSF things are moved into ember_osf_web

File: website/quickfiles/views.py [hunk 12]
```diff
@@ -0,0 +1,14 @@
...
+quickfiles_dir = os.path.abspath(os.path.join(os.getcwd(), EXTERNAL_EMBER_APPS['quickfiles']['path']))
```
Comment (jamescdavis) [inline website/quickfiles/views.py]: s/quickfiles/ember_osf_web/g

File: website/quickfiles/views.py [hunk 13]
```diff
@@ -0,0 +1,14 @@
...
+
+def use_ember_app(**kwargs):
+    if PROXY_EMBER_APPS:
+        resp = requests.get(EXTERNAL_EMBER_APPS['quickfiles']['server'], stream=True)
```
Comment (jamescdavis) [inline website/quickfiles/views.py]: s/quickfiles/ember_osf_web/

File: website/quickfiles/views.py [hunk 14]
```diff
@@ -0,0 +1,14 @@
...
+        return Response(stream_with_context(resp.iter_content()), resp.status_code)
+    return send_from_directory(quickfiles_dir, 'index.html')
```
Comment (jamescdavis) [inline website/quickfiles/views.py]: s/quickfiles/ember_osf_web/

**Comment by hmoco** [comment]
@jamescdavis ready for another code review pass!

**Comment by hmoco** [comment]
@jamescdavis @sloria tests pass locally, but fail consistently (4/4 after clearing cache) on travis. Any suggestions? This is the (seemingly unrelated) one test that is failing:
```
___________ ERROR collecting api_tests/test/views/test_throttling.py ___________
ImportError while importing test module '/home/travis/build/hmoco/osf.io/api_tests/test/views/test_throttling.py'.
Original error message:
'No module named test_throttling'
Make sure your test modules/packages have valid Python names.
```

**Comment by sloria** [comment]
@hmoco Try adding a blank `api_tests/test/__init__.py` file.

**Comment by hmoco** [comment]
@sloria, that seems to have done it. Weird. Thanks!

File: website/routes.py [hunk 15]
```diff
@@ -296,6 +296,18 @@ def make_url_map(app):
                 ),
             ], prefix='/' + prefix)
 
+            if value.get('has_additional_routes', False):
+                import imp
+                app_views = imp.load_source('views', 'website/' + value['url'] + '/views.py')
+                process_rules(app, [
+                    Rule(
+                        [route for route in app_views.routes],
```
[the remark this answers is not in the document]
Comment (hmoco) [reply website/routes.py]: :/

[the remark this answers is not in the document]
Comment (hmoco) [reply website/routes.py]: This is what happens when I don't write python in a while and wanna use my limited toolset to its full ( ;) ) potential

File: addons/base/views.py [hunk 16]
```diff
@@ -692,9 +691,9 @@ def addon_view_or_download_file(auth, path, provider, **kwargs):
     return addon_view_file(auth, node, file_node, version)
 
 
-def addon_view_or_download_quickfile(**kwargs):
+def addon_view_or_download_file_by_fid(**kwargs):
```
Comment (jamescdavis) [inline addons/base/views.py]: We can remove this since the front-end will no longer use the file_redirect route

**Comment by jamescdavis** [comment]
@sloria and/or @brianjgeiger can this go into the next OSF release?

**Comment by hmoco** [comment]
🔥

**Pull request merged**